src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -251,7 +256,36 @@ private static void AddRouting(OrchardCoreBuilder builder)
 
                 collection.AddRouting();
             },
-            order: int.MinValue + 100);
+            order: Int32.MinValue + 100);
+        }
+
+        /// <summary>
+        /// Configures ApiExplorer at the tenant level using <see cref="Endpoint.Metadata"/>.
+        /// </summary>
+        private static void AddEndpointsApiExplorer(OrchardCoreBuilder builder)
+        {
+            // 'AddEndpointsApiExplorer()' is called by the host.
+
+            builder.ConfigureServices(collection =>
+            {
+                // Remove the related host singletons as they are not tenant aware.
+                var descriptorsToRemove = collection
+                    .Where(sd => sd is ClonedSingletonDescriptor &&
+                        (sd.ServiceType == typeof(IActionDescriptorCollectionProvider) ||
+                        sd.ServiceType == typeof(IApiDescriptionGroupCollectionProvider)))
+                    .ToArray();
+
+                foreach (var descriptor in descriptorsToRemove)
+                {
+                    collection.Remove(descriptor);
+                }
+
+#if NET6_0_OR_GREATER
+                // Configure ApiExplorer at the tenant level.
+                collection.AddEndpointsApiExplorer();
```

Comment:
This is enabling the api explorer by default. Can't we just do that if a service was previously registered in the `collection`? This way we only enable it at the tenant level if it was invoked by the main Startup class of the user's app.

Reply:
I did this because it will be called automatically by the web host, just after `AddRouting()`. It will register the 2 default singletons of `IActionDescriptorCollectionProvider` and `IApiDescriptionGroupCollectionProvider`, see the comment below => `Try to add default services in case MVC services aren't added.`

    public static IServiceCollection AddEndpointsApiExplorer(this IServiceCollection services)
    {
      // Try to add default services in case MVC services aren't added.
      services.TryAddSingleton<IActionDescriptorCollectionProvider, DefaultActionDescriptorCollectionProvider>();
      services.TryAddSingleton<IApiDescriptionGroupCollectionProvider, ApiDescriptionGroupCollectionProvider>();


We can still register a custom implementation at the host level, will be removed here but just at the tenant level. Then it's up to the user to register a custom implementation at the tenant level by using a module startup or our app helper. Then, as we do for `AddRouting()`, here we call `AddEndpointsApiExplorer()` that adds the defaults singletons.

Reply:
This seems too special, can't we do something like make a fake service collection and add a bunch of services we don't want to chain back to the host application then special case those?

Reply:
@davidfowl Thanks for the feedback, yes I thought about it

First some contextual info, yes we build a child container and pipeline for each tenant, a tenant is composed of modules that have their own startup so that they can collaborate to build a tenant container and pipeline, and we have helpers allowing the application to register some delegates to do the same collaboration at the tenant level.

- Yes, the first time we filtered some services we did it directly when we create a tenant child container, it was for the host `IStartupFilter` to prevent them to re-add middlewares when we build a given tenant pipeline. https://github.com/OrchardCMS/OrchardCore/blob/9d522b4895f320949e0c97992b03a14245fac0dd/src/OrchardCore/OrchardCore/Shell/Builders/Extensions/ServiceProviderExtensions.cs#L23

- Most of the time we only need to re-register / remove some services, **for now we did it step by step on each new issue related to a specific `AddSomething()`**, this in different modules startups or with helpers from the app.

- The problem is mainly related to host singletons / options, particularly when they hold global collections. One problem is when a service is registered with an implementation instance whose type is not public, in that case we use a fake collection on which we do the same `AddSomething()` to retrieve the implementation types.

- E.g. when we migrated from 2.0 to 3.0 with the routing services, since 3.0 the route `IOptions` is mutated each time `UseEndpoints()` is called, but our tenant pipelines may be built and then call `UseEndpoints()` concurrently, so we needed isolated ones per tenant. See the following and other `AddSomething()` in the same file. https://github.com/OrchardCMS/OrchardCore/blob/9d522b4895f320949e0c97992b03a14245fac0dd/src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs#L242

Just to say that we already did it in many places in modules or with helpers from the app and for now most of the time related to different specific `AddSomething()`, that's why here I also did it specifically. But yes, maybe better to do it in a more central place e.g. when we first create a given tenant child container.

That said we may not be able to do it in only one place, the library where we create tenant containers may not reference some needed packages to use some service types and `AddSomething()` (e.g. with `OpenIddict`). As soon as I have time I will focus on it and will try to filter services in a a more centralized way, thanks for the suggestion.

Reply:
> That said we may not be able to do it in only one place, the library where we create tenant containers may not reference some needed packages to use some service types and AddSomething() (e.g. with OpenIddict). 

If you want to discuss about potential ways to improve things in OpenIddict, let me know.

Reply:
@kevinchalet Okay cool!

When I will have time to work on it, I will let you know, thanks.